internal/btf/ext_info.go

```diff
@@ -98,156 +96,268 @@ func parseBTFExtCoreHeader(r io.Reader, bo binary.ByteOrder, extHeader *btfExtHe
+		return "", nil, fmt.Errorf("get section name: %w", err)
 	}
 
-	buf := internal.NewBufferedSectionReader(sec, extHeader.funcInfoStart(), int64(extHeader.FuncInfoLen))
-	funcInfo, err = parseExtInfo(buf, bo, strings)
-	if err != nil {
-		return nil, nil, nil, fmt.Errorf("function info: %w", err)
+	if infoHeader.NumInfo == 0 {
+		return "", nil, fmt.Errorf("section %s has zero records", secName)
 	}
 
-	buf = internal.NewBufferedSectionReader(sec, extHeader.lineInfoStart(), int64(extHeader.LineInfoLen))
-	lineInfo, err = parseExtInfo(buf, bo, strings)
-	if err != nil {
-		return nil, nil, nil, fmt.Errorf("line info: %w", err)
+	return secName, &infoHeader, nil
+}
+
+// parseExtInfoRecordSize parses the uint32 at the beginning of a func_infos
+// or line_infos segment that describes the length of all extInfoRecords in
+// that segment.
+func parseExtInfoRecordSize(r io.Reader, bo binary.ByteOrder) (uint32, error) {
+	const maxRecordSize = 256
+
+	var recordSize uint32
+	if err := binary.Read(r, bo, &recordSize); err != nil {
+		return 0, fmt.Errorf("can't read record size: %v", err)
 	}
 
-	if coreHeader.CoreReloOff > 0 && coreHeader.CoreReloLen > 0 {
-		buf = internal.NewBufferedSectionReader(sec, extHeader.coreReloStart(coreHeader), int64(coreHeader.CoreReloLen))
-		relos, err = parseExtInfoRelos(buf, bo, strings)
-		if err != nil {
-			return nil, nil, nil, fmt.Errorf("CO-RE relocation info: %w", err)
-		}
+	if recordSize < 4 {
+		// Need at least InsnOff worth of bytes per record.
+		return 0, errors.New("record size too short")
+	}
+	if recordSize > maxRecordSize {
+		return 0, fmt.Errorf("record size %v exceeds %v", recordSize, maxRecordSize)
 	}
 
-	return funcInfo, lineInfo, relos, nil
+	return recordSize, nil
 }
 
-type btfExtInfoSec struct {
-	SecNameOff uint32
-	NumInfo    uint32
+// FuncInfo represents the location and type ID of a function in a BPF ELF.
+type FuncInfo struct {
+	InsnOff uint32
+	TypeID  TypeID
 }
 
-type extInfoRecord struct {
-	InsnOff uint64
-	Opaque  []byte
+// Name looks up the FuncInfo's corresponding function name in the given spec.
+func (fi FuncInfo) Name(spec *Spec) (string, error) {
+	// Look up function name based on type ID.
+	typ, err := spec.TypeByID(fi.TypeID)
+	if err != nil {
+		return "", fmt.Errorf("looking up type by ID: %w", err)
+	}
+	if _, ok := typ.(*Func); !ok {
+		return "", fmt.Errorf("type ID %d is a %T, but expected a Func", fi.TypeID, typ)
+	}
+
+	// C doesn't have anonymous functions, but check just in case.
+	name, _ := typ.TypeName()
+	if name == "" {
+		return "", fmt.Errorf("Func with type ID %d doesn't have a name", fi.TypeID)
+	}
+
+	return name, nil
 }
 
-type extInfo struct {
-	byteOrder  binary.ByteOrder
-	recordSize uint32
-	records    []extInfoRecord
+// Marshal writes the binary representation of the FuncInfo to w.
+// The instruction offset is converted from bytes to instructions.
+func (fi FuncInfo) Marshal(w io.Writer, offset uint64) error {
```

Comment:
Planning to do this when getting rid of `btf.Program`, since implementing `FuncInfos.Marshal()` requires all FuncInfo's offsets being correctly prepopulated. The only good source of truth for marshaling a `FuncInfos` from is a `[]*(ebpf/btf).Program` (like the output of `Program.flatten()`), since they're order-dependent and the offsets need to correspond to the Programs' positions in the instruction stream.